plutus-core/plutus-core/src/PlutusCore/Flat.hs

```diff
@@ -85,6 +88,16 @@ implementations for them (if they have any constructors reserved for future use)
 By default, Flat does not use any space to serialise `()`.
 -}
 
+-- | For deriving 'Flat' instances via 'Serialize'.
+newtype AsSerialize a = AsSerialize
+    { unAsSerialize :: a
+    } deriving newtype (Serialise)
+
+instance Serialise a => Flat (AsSerialize a) where
+    encode = encode . serialise
+    decode = deserialise <$> decode
```

Comment:
This is a bit unsafe:
- `decode` can throw `DeserializeException`s from `flat`, but they get caught and `unflat` returns an `Either`
- `deserialise` also throws, but it throws exceptions from `serialise` which won't get caught by `flat`.

So if you try and `unflat` an invalid `Data` literal you'll get an exception even though `unflat` suggests it should catch everything.

It think the solution is just to use `deserialiseOrFail` and then throw the right kind of exception in the error case. Could just use the `MonadFail` instance, I think.

Reply:
> It think the solution is just to use deserialiseOrFail and then throw the right kind of exception in the error case. Could just use the MonadFail instance, I think.

Did that, thanks.